corehq/ex-submodules/couchforms/models.py

```diff
@@ -346,27 +346,56 @@ def _meta_to_json(meta):
     def xml_md5(self):
         return hashlib.md5(self.get_xml().encode('utf-8')).hexdigest()
 
-    def archive(self, user_id=None):
-        if self.is_archived:
-            return
-        self.doc_type = "XFormArchived"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='archive',
-        ))
-        self.save()
-        xform_archived.send(sender="couchforms", xform=self)
+    def archive(self, user_id=None, retry_archive=False):
+        if not retry_archive:
+            if self.is_archived:
+                return
+            # If this archive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True) as archive_stub:
+            self.doc_type = "XFormArchived"
+
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='archive',
+            ))
+            archive_stub.archive_history_updated()
+            self.save()
+            xform_archived.send(sender="couchforms", xform=self)
+
+    def unarchive(self, user_id=None, retry_archive=False):
+        if not retry_archive:
+            if not self.is_archived:
+                return
+            # If this unarchive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False) as archive_stub:
+            self.doc_type = "XFormInstance"
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='unarchive',
+            ))
+            archive_stub.archive_history_updated()
```

Comment:
should be after the save

Reply:
https://github.com/dimagi/commcare-hq/pull/22470/commits/a2a9850d9000a53e1294aee61d649dd23ca55d7f